Cellxgene schema cli v 1 0 0 uns/var/obs/obsm/X validator

pablo-gar opened the pull request:
Second iteration of `cellxgene-schema` v1.0.0. Adds:

- Implements CLI:

	- `cellxgene-schema validate`
	
- Adds `validate.py` with the following classes:

	- `Validator` handles validation of an AnnData object
	- `LabelWriter` handles addition of metadata label as defined in the schema
	- `__main__` is the entry point of CLI

- Expands testing:
	- Adds tests for the classes shown above
	- Adds tests for the CLI entry point.
	- Adds minimal h5ad files and anndata objects under test fixturtes


Example of important additions (CLI):

```bash
# Valid h5ad 
cellxgene-schema validate example_valid.h5ad

# Inject label metadata as indicated in schema
cellxgene-schema validate -o example_with_labels.h5ad example_valid.h5ad


cellxgene-schema validate example_invalid_CL.h5ad
#ERROR: 'PATO:0000461' in 'cell_type_ontology_term_id' is not a valid ontology term id of 'CL'
#ERROR: 'CL:12312' in 'cell_type_ontology_term_id' is not a valid ontology term id of 'CL'

cellxgene-schema validate example_invalid_uns.h5ad
#ERROR: '1' in 'uns['title']' is not valid, it must be a string.
#ERROR: Value 'batchD' of list 'batch_condition' is not a column in 'adata.obs'
#ERROR: Value 'batchE' of list 'batch_condition' is not a column in 'adata.obs'
#ERROR: 'X_PCA_1' in 'uns['default_embedding']' is not valid, it must be a key of 'adata.obsm'.
#ERROR: '1' in 'uns['X_normalization']' is not valid, it must be a string.
#ERROR: 'CPM' in 'uns['X_approximate_distribution']' is not valid. Allowed terms: ['count', 'normal'].

cellxgene-schema validate example_invalid_obsm.h5ad
#ERROR: All embeddings must have as many rows as cells, and at least two columns.'adata.obsm['X_umap']' has shape of '(320, 1)'


cellxgene-schema validate example_invalid_raw.h5ad
#ERROR: Matrix 'X' seems to be the raw matrix but not all of its values are integers.


cellxgene-schema validate example_invalid_reserved_terms.h5ad
#ERROR: Add labels error: Column 'cell_type' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'assay' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'disease' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'organism' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'sex' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'tissue' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'ethnicity' is a reserved column name of 'obs'. Remove it from h5ad and try again.
#ERROR: Add labels error: Column 'development_stage' is a reserved column name of 'obs'. Remove it from h5ad and try again.
```

Addresses:
#83 
#84 
#85  
#86  
#87  
#88  
#89  
#90  
#91 
#94  
#95  
#96 
#97  
#98 
#109

atolopko-czi commented on cellxgene_schema_cli/tests/fixtures/examples_validate.py:
```diff
@@ -0,0 +1,205 @@
+    ],
+)
+
+obs_expected = pd.DataFrame(
```
I'm confused by the `expected` term here, since it's not being used directly in an assertion; only used to created to init `adata_with_labels`, below. Does `expected` mean something not related to test assertions?

atolopko-czi commented on cellxgene_schema_cli/tests/fixtures/examples_validate.py:
```diff
@@ -0,0 +1,205 @@
+    var=pd.concat([good_var, var_expected], axis=1),
+    uns=good_uns,
+    obsm=good_obsm
+)
```
There's a lot to grok in this file to understand what a specific test is using as its test setup state.  Not sure how much of these objects are globally used throughout tests, but if they can be segregated a bit and moved within a scope that makes them visible only to the groups tests they concern, it can help the reader.

atolopko-czi commented on cellxgene_schema_cli/tests/fixtures/examples_validate.py:
```diff
@@ -0,0 +1,205 @@
+    "example_invalid_ethnicity.h5ad",
+    "example_invalid_development_stage.h5ad",
+    "example_invalid_uns.h5ad",
+]
```
As a test reader, it's opaque as to what the actual invalid data item is, other than from the name. So from the standpoint of verifying that a test is accurate, I'm wondering if this fixture-based test setup strategy could be replaced by a more dynamic one: E.g., starting with a single _valid_ AnnData object--which is constructed programmatically so the reader can easily see its contents in the test code--and then for each error test case, mutating the one item that should cause a validation to be triggered.

atolopko-czi commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,408 @@
+            self.assertTrue(self.validator.errors)
+
+
+class TestColumnValidation(unittest.TestCase):
```
suggest `TestObsAndVarColumnValidation` for quick identification of the scope of this class.

atolopko-czi commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,408 @@
+        )
+        self.assertTrue(self.validator.errors)
+
+    def test_ontology_columns(self):
```
This test is a schema-level test, while the ones above seem to be lower-level column validation logic tests that are not tied to a schema. Thinking it might good (throughout all tests) to separate those that are testing schema constraints vs the lower level implementation of schema-agnostic constraints.  Could even separate into different test files, so it's clear that, say, one file is all about testing schema 2.0.0 specifically.

atolopko-czi commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,408 @@
+
+    def test_validate_bad_obsm(self):
+
+        # Wrong dimension, delete one column
```
could move comment into the method name, and del the comment

atolopko-czi commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,408 @@
+
+    def test_schema_defintion(self):
+        """
+        Tests that the definition of schema is correct
```
```suggestion
        Tests that the definition of schema is well-formed
```

I'm worried that "correct" might imply that it meets the written 2.0.0 spec, while, I believe that this test is just checking the structure of the schema file can be parsed. Is that correct?

atolopko-czi commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,408 @@
+from scipy import sparse
+
+
+class TestFieldValidation(unittest.TestCase):
```
What does `Field` mean in this case? These tests are not testing validation of  specific AnnData object, but rather the schema file itself, and some unit tests for `_validate_curie`.  Is this is cohesive grouping of tests? Maybe the class be renamed for clarity.

atolopko-czi reviewed (commented):
I've mostly spent time just reviewing the overalls structure tests and made some comments from a very initial, naive perspective!  Coming at this as a new reader of the tests, my comments  are just about clarifying the intent of some tests and how state setup is used and shared across various tests. I haven't yet fully visualized how well the tests provide coverage of the schema specification, and realize that coverage might quite comprehensive as is.  But in trying to determine the level of coverage, I was thinking it could be worth considering restructuring so that the tests are more specification-oriented. That is, for each "MUST", "MUST NOT" declaration in the spec, there's a corresponding test with a test name that mirrors a specification "statement". And then each test would do all of its own test state setup (and not share any setup state), making it clear to the test reader what specific single item is being made invalid and that should cause a validation error.  It might be good have assertions check the actual error message too, rather than just the existence of _some_ error.  

As I'm going to be away for PTO, please ignore anything I've written since I won't be able to dive into deeper discussion. :) Happy to revisit later on, if helpful!

pablo-gar commented:
Thanks for report Jenny see my inline responses


> * **uns**: 'X_approximate_distribution is in uns, but value is None' results in no error, but should result in logging an error, as value must be _str_

This is an Anndata-related issue.  If anything in `uns` is assigned `None` and then proceed to save the h5ad, that key in uns won’t actually get written, thus leading to the described behavior. 

> * **uns**: 'X_normalization is 'none', but there is a matrix in raw.X' results in no error, but should result in logging an error, as 'none' should be reserved for datasets without a normalized matrix

Corrected in edc81ad3929692bbe79d924cafadb88a1b0e13c4

> * **uns**: 'title leading/trailing white spaces, double spaces' results in no error, but should result in logging an error as we have found leading/trailing white spaces, double spaces in fields that can eventually cause issues (let me know if you want to discuss this)

We are having a conversation over slack to decide whether we will check this.

> * **obsm**: ~"have key not start with 'X_'" results in no error, but should result in logging an error as embeddings must be prefixed with 'X_'~ obsm containing only 'umap' and 'tsne' as keys for the embeddings results in no error, but should result in logging an error saying it should contain at least 1 embedding starting with 'X_'

Corrected in edc81ad3929692bbe79d924cafadb88a1b0e13c4

> * **obsm**: 'np.ndarray of shape (n, m), where m == 1' is causing validator to throw an error ('IndexError: tuple index out of range')

Corrected in edc81ad3929692bbe79d924cafadb88a1b0e13c4

pablo-gar commented:
f292e4e782596ae06703029470ab047f19513e7b adds the new requirement in the schema for development stage:

> Otherwise, for all other organisms this MUST be the most accurate child of UBERON:0000105 for life cycle stage, excluding UBERON:0000071 for death stage.

jychien commented:
Hi Pablo,
I've found a few issues with var, and will hold off on raw.var validation until after the validator is ready for that:
- **var**: reserved column 'feature_name' must not be annotated and validation results in no error
- **var**: reserved column 'feature_reference' must not be annotated and validation results in no error
- **var**: 'ENSG00000060303' is a gene in gencode.v32.primary_assembly.annotation.gtf, but is a processed_pseudogene, so is not in cellranger 2020-A reference. My expectation is that this is still a valid gene, but it is flagged by the validator as 'not a valid feature IDnot a valid feature ID'.

pablo-gar commented:
Thanks for the report @jychien. Everything has been fixed and validation of `raw.var` has been added (63c516e5a404c279d5a718cb9dc234db49203e16) , please feel free to continue your checks using the latest commit.

Detailed fixes:

> * **var**: reserved column 'feature_name' must not be annotated and validation results in no error

Since this comes from the index the validator was not looking for this, fixed in cb8accee4227c6b2987491cd0405d11ede8990c6

> * **var**: reserved column 'feature_reference' must not be annotated and validation results in no error

Same as above

> * **var**: 'ENSG00000060303' is a gene in gencode.v32.primary_assembly.annotation.gtf, but is a processed_pseudogene, so is not in cellranger 2020-A reference. My expectation is that this is still a valid gene, but it is flagged by the validator as 'not a valid feature IDnot a valid feature ID'.

I have replaced the 10X filtered GTF files for the Gencode files in 66e30ec332e83da9f70adedb6e39919170815a1b

pablo-gar commented:
@atolopko-czi @MDunitz 

Tests have refactored:

- `test_ontolgy.py` Tests for internal functions of the `ontolgy.OntologyChecker` and `ontology.GeneChecker` classes.
- `test_validate.py` Tests for internal functions of the `validate.Validator` and `validate.LabelWriter` classes.
- `test_schema_compliance`. Follows schema definition and test for failure cases for each MUST/SHOULD in the schema (only SHOULDs that can be checked are added). It directly tests for the expected error or warning.
- `fixtures/examples_validate.py` Has been heavily commented for improved reading.

atolopko-czi commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'ethnicity_ontology_term_id' is not a "
+                "valid value of 'ethnicity_ontology_term_id'"
```
should the err msg be `must be 'na' for non-human organism 'NCBITaxon:10090'`?

pablo-gar replied:
Fixed in f68682ff4c918f36b5fae8f4fab1d2b98201caf4

atolopko-czi commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        """
+
+        # If organism_ontolology_term_id is "NCBITaxon:9606" for Homo sapiens,
+        # this MUST be either a HANCESTRO term or "unknown" if unavailable.
```
comment seems to be a copy&paste err. fix or rm?

pablo-gar replied:
I wanted to emphasize the code directly below corresponds that schema spec

atolopko-czi commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'organism_ontology_term_id' is not a valid "
+                "ontology term id of 'NCBITaxon'"
```
```suggestion
                "child ontology term id of NCBITaxon:33208 for Metazoa'"
```

Is the `NCBITaxon:33208` term itself allowed or must the term strictly be a child?

brianraymor replied:
Per the schema - This MUST be a child of NCBITaxon:33208 for _Metazoa_.

pablo-gar replied:
Fixed in f68682ff4c918f36b5fae8f4fab1d2b98201caf4

atolopko-czi commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'sex_ontology_term_id' is "
+                "not a valid ontology term id of 'PATO'"
```
```suggestion
                "neither a valid ontology term id of 'PATO' nor the value 'unknown'"
```

pablo-gar replied:
Fixed in f68682ff4c918f36b5fae8f4fab1d2b98201caf4

atolopko-czi commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+                that do not capture the tissue of origin (e.g. In the case of CD3+ kidney cells, use "UBERON:0002113"
+                for kidney instead of "CL:000084" for T cell. However, in the case of EPCAM+ cervical cells,
+                use "CL:000066" for epithelial cell of the cervix.)
+        """
```
There's a lot of constraints here, but it seems they cannot be programmatically verified here, since we don't know the type of the biological sample, right?

brianraymor replied:
[`tissue_ontology_term_id`](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/2.0.0/corpora_schema.md) started with a simple MUST for UBERON and then evolved over time for cases that cannot be programmatically verified. 

The final _Enriched, Sorted, or Isolated Cells from a Tissue_ case is the proverbial straw. It has become more like the STRONGLY RECOMMENDED that we use for `development_stage_ontology_term_id`. 

We should probably rewrite to clarify:

1  curie MUST BE UBERON or CL - with STRONGLY RECOMMENDED details for the cases but only enforcing the ontology selection
2. MUST BE UBERON (organoid) - if the curator annotates for organoids
3. MUST BE CL (cell culture) - if the curator annotates for cell cultures

@ambrosejcarr ^^^

pablo-gar replied:
Correct

brianraymor replied:
Addressing in #131

pablo-gar replied:
Finished in #131

atolopko-czi reviewed (commented):
The re-org of the tests is a great improvement! The inclusion of the explicit error msgs makes this test code much more robust and verifiable. Thanks for all those changes!

The singular "happy path" test in `TestValidAnndata` is a nice, singular test to have. One consideration is that it detaches the "valid" cases from the "invalid" cases in the test code. I realized later on that some schema constraints have multiple valid cases, such as `MUST be either a HANCESTRO term or "unknown" if unavailable` are in fact properly tested by `TestValidAnndata`, so that's great. It just took me (the reader) more time to confirm that. Ultimately, we could move each of these valid cases together with the invalid case test methods, thus grouping valid and invalid tests by schema requirement. For now, it's probably just worth adding additional comment to `TestValidAnndata` to indicate that _all_ the valid cases are being exercised there, and maybe repeat that in the comments for the relevant invalid tests when there are multiple valid cases. This also makes me think we could rename most of the test methods to `test_[invalid_]*` to drive home it's just exercising the invalid cases (there are some testing the "extra" valid cases as well).

File renaming suggestions:
* `test_schema_compliance.py` to `test_schema_data_validation.py`
* `test_validate.py` to `test_schema_definition.py`
Would also add comments in these test files to explain the difference so the reader can orient herself more quickly. 

Wondering if err msgs could actually include a link (or other ref code) that points to the specific schema 2.0 requirement that is violated. Maybe not necessary if the err msg includes enough info, but thought I'd mention.

In addition to just reading through the PR, I read through the schema 2.0.0 doc itself, and to ensure the the unit tests are validating the schema as stated. However, I've only completed this up to the [`var and raw.var`](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/2.0.0/corpora_schema.md#var-and-rawvar-gene-metadata) section. So more to go.  I'm going to submit my review as `Comment` for now, in case there is feedback you want to address sooner than later. I will finish up this PR on Wed!

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0009310' in 'assay_ontology_term_id' is a deprecated term id of 'EFO'",
+                "ERROR: 'EFO:0009310' in 'assay_ontology_term_id' is not a children term id "
```
As of now, we have control of the error messages. @brianraymor may want to do a pass in the final review.

brianraymor replied:
> also, might we put the ontology version in the error to to help curators debug?

@jahilton and @jychien - would that be helpful or _noise_?

jahilton replied:
I don't foresee the ontology version being particularly helpful.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        """
+        development_stage_ontology_term_id categorical with str categories. If unavailable, this MUST be "unknown".
+        If organism_ontolology_term_id is "NCBITaxon:9606" for Homo sapiens,
+        this MUST be the most accurate HsapDv term.
```
It means the curator must used most exact/precise term that fits their data.

the MUST relates to being a child HsapDv -- we cannot test for accuracy of the term

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'disease_ontology_term_id' is not a "
+                "valid ontology term id of 'MONDO, PATO'"
```
The error msg is accurate but not specific. This msg is produced as part of programatic check, adding an exception for this case may make the code fragile. I'd like to know your thought after your review `validate.py`

However there's a specific error msg when  a PATO term is used but it is not 'PATO:0000461'. I have just added an extra test for that.

atolopko-czi replied:
Ah great, that should be sufficient.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'tissue_ontology_term_id' is not a "
+                "valid ontology term id of 'UBERON, CL'"
```
Similar as above the error msg is accurate but not to the highest precision as this is part of a programatic check. Let me know what you think after you review `validate.py`

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        Otherwise, this MUST be False.
+        """
+
+        # Duplicate 1st row in var and assigned to 2nd
```
I don't think so, unless you are referring strictly to "duplicate"

atolopko-czi replied:
ok, it just seemed like that comment for was `new_index[1] = new_index[0]` line 417, above.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            ],
+        )
+
+    def test_X_normalization_not_raw(self):
```
I think I can go in either. I left it here because the test relates to the value un uns['X_normalization'] as well

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: Matrix 'X' seems to be the raw matrix but not all of "
+                "its values are integers."
```
In the code the test is actually done using fractional parts, so floats are allowed as long as they do not have decimal values (i.e. referring to mathematical integers)

@brianraymor any comments on whether to include the kind of test that is performed in the actual schema?

brianraymor replied:
If our requirements are more restrictive than AnnData, then it would be helpful to define the requirements for the values in the schema.

pablo-gar replied:
Our schema requirements are in line with the checks, but I agree that it may not be obvious to the user that an "integer" check is being done. I don't think we should include it in the schema as it may add more noise to an already busy section. If this check causes confusion in the future we can add some clarifying text to the schema.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_obsm_values_at_least_one_X(self):
+
+        """
+        At least one key for the embedding MUST be prefixed with "X_"
```
Yes, It is acceptable to have some embeddings that are not prefixed with `X_`, any embedding that does not have `X_` will not be used by cellxgene and therefore some of the checks are not done on those. The only check that is done in all embeddings is the numpy array check.

pablo-gar replied:
@brianraymor any thoughts on including the suggestion in the schema?

brianraymor replied:
_everything is permitted_ if it's not prohibited in the schema. How would this information help curators?

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_obsm_shape(self):
+
+        """
+        Curators MUST annotate one or more two-dimensional (m >= 2) embeddings
```
embeddings in single-cell data are used for other purposes aside visualization, e.g. PCs from PCA can be used for both visualization and  analyses: the first 2 PCs can used to visualize the data whereas the rest of the PCs can be used to perform some sort of analysis on the data that users who download it may want to have access to.

atolopko-czi commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+                        "the normalization used for 'X'."
+                    )
+
+                if self._get_raw_x_loc() == "raw.X" and not self._is_raw():
```
Should this check also be made when `X_norm` _is_ specified?  It's currently nested under the case where `X_norm == 'none'`.

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

atolopko-czi commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        Checks if elements in the specified column of the component (e.g. 'assay_ontology_term_id' of 'adata.obs') are
+        children of the given ancestors.
+
+        Ancestors checks are inclusive, meaning that a value it's its own ancestor as well.
```
```suggestion
        Ancestors checks are inclusive, meaning that a value is its own ancestor as well.
```

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

atolopko-czi commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        else:
+            return "X"
+
+    def _is_raw(self, n_values_to_check: int = 5000, force: bool = False) -> bool:
```
any rationale to the 5000 value?   if checking more than this is a performance concern, worth a comment, and could rename to `max_values_to_check`

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

atolopko-czi reviewed (approved):
Finished checking that schema doc has comprehensive unit tests. Found only a few cases where valid values are possibly not being verified by a test.

Perused `validate.py`, and looks good, with a few minor refactoring suggestions.

It seems that there's a good bit a complexity around allowing _either_ raw.X or X to contain raw data and the validation around checking raw values and the X_norm annotation. (I can understand why the schema allows for this, however.) I haven't quite determined if there's a way to simplify any of the code, but will come back to it if I form any ideas. 

No clear blockers, so approving, but do have a look through my final set of comments, in case anything is worth changing.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_raw_existance(self):
+
+        """
+        Except for ATAC-seq and methylation data, raw data is REQUIRED
```
Thanks for the suggestion, I prefer not linking out as the schema links can change based on updates or editorial changes

brianraymor replied:
This would be very cool, but I agree with @pablo-gar - it would be a maintenance **bear**.

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'EFO:0000001' in 'development_stage_ontology_term_id' is "
+                "not a valid ontology term id of 'HsapDv'"
```
Based on feedback from curators this would not be as useful

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            ["ERROR: Column 'index' in dataframe 'var' is not unique."],
```
This is a good suggestion but out of scope IMO and simple enough for the curator to debug.

brianraymor replied:
I'm a little confused about all the comments that are referencing test script _messaging_ versus the output from validate.py like `"Column '{column_name}' in dataframe '{df_name}' is not unique."`

Maybe I need more caffeine ;-)

jychien commented:
Hi @pablo-gar, I'm still going through test cases for X and raw.X, but I thought I'd list these initial issues, as one of the error messages is coming up on more than 1 occasion (which was the one I had slacked you about). In general, I'm am creating test cases where I am **not** assuming that 'X_normalization' was filled accurately, so the validator should flag datasets that are missing the raw layer when applicable:
-  **X/raw.X**: ATAC-seq dataset with only normalized in X (should be okay), but results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF180GFI_v4.h5ad)
- **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, and X_normalization = 'LogNormalization' results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF751GQY_v4_2.h5ad)
- **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, X_normalization = none results in no logging error (test file = LATDF751GQY_v4_1.h5ad)
- **X/raw.X**: have X and raw.X be different shapes results in no logging error (test file = LATDF516GAW_v4_2.h5ad)

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        self.validator.adata.obs["assay_ontology_term_id"] = "EFO:0010891"
+        self.validator.validate_adata()
+        print("HOLAaaa", self.validator.errors)
+        self.assertEqual(self.validator.errors, [])
```
I just added some validation and tests, will push them shortly.

All of the checks to the layers were the most complex ones, thanks for this catch

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+            if raw_loc == "raw.X":
+                x = self.adata.raw.X
+            else:
+                x = self.adata.X
```
this is the only place where this execution is done, I'm gonna leave it as is for now instead of creating a new function

jahilton commented:
Got this error message when for ethnicity_ontology_term_id:unknown for non-human organism
>ERROR: 'unknown' in 'ethnicity_ontology_term_id' is not a valid value of 'ethnicity_ontology_term_id'

I would find it very useful to have the organism condition specified in the message in case the organism is the root cause of the error, something like...
>ERROR: 'unknown' in 'ethnicity_ontology_term_id' is not a valid value of 'ethnicity_ontology_term_id' for 'organism' of 'NCBITaxon:10090 [Mus musculus]'

brianraymor commented:
In scripts/readme.md:

> Then the following will command download and process the reference files, and then update the respective package files under `cellxgene_schema_cli/cellxgene_schema/ontology_files/`

should be:

Then the following command will download

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+
+        columns = [
+            "feature_is_filtered",
+        ]
```
`feature_id` is the index, I just added `feature_biotype`

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        return self.is_valid
+
+
+class LabelWriter:
```
Can you expand on why it's worth moving it? 

`H5adLabelAppender` is initialized with a`Validator` instance. And both classes are used in the function `validate` which is the entry point of the CLI.

They all are in the same file for convenience.

atolopko-czi replied:
Primarily to reduce the size of the file and improve readability.  As a reader of the code, scanning the files/packages is usually my first introduction to the structure of the code and the various components. (It's unfortunate IDE are still file-oriented, rather than class-oriented!).  Anyway, quite minor, no strong reason to change it.

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        # Fetches schema def from anndata if schema version is not found in AnnData, this fails
+        self._set_schema_def()
+
+        if not self.errors:
```
This if is because `self._set_schema_def()` _can_ add errors _if_ the schema version is not supported. I'm not sure I follow your suggestion.

atolopko-czi replied:
ah, I see. I figured it was trying to avoid performing checks redundantly if the method was called multiple times; I didn't consider the side-effect of `_set_schema_def()`.

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+            if "add_labels" in index_def:
+                self._add_column(component, "index", index_def)
+
+    def write_labels(self, add_labels_file: str):
```
Yeah I thought about it but I favored consistency of error handling between the `Validator` and the `LabelWriter` classes. If there's a strong reason to change it to this model I'd be happy to do so (e.g. better error handling in the backend?)

atolopko-czi replied:
no strong reason; was just considering the "single-responsibility principle"

jychien commented:
I am currently testing a sample h5ad that has gene names as feature ID. The gene names, as expected, do fail, but the error logging is a little confusing. For example:
- ERROR: Could not infer organism from feature ID 'AURKAIP1' in 'var', make sure it is a valid ID
- ERROR: 'ERCC3' is not a valid feature ID in 'var'
- ERROR: Column 'feature_biotype' in dataframe 'var' contains invalid values '['gene']'. Values must be one of ['spike-in'].

For the first error, I am expecting more straight forward telling the curator that it is not a valid feature ID. For the second, 'ERCC3' is actually a gene name, while actual spike-in identifiers have a hyphen, such as in 'ERCC-00170'. So, the third ERROR would be confusing as it is not a spike-in.

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+    return yaml.load(open(path), Loader=yaml.FullLoader)
+
+
+def _get_organism_from_feature_id(
```
Done for `_get_organism_from_feature_id`

`_curie_remove_suffix` won't be moved as it uses an `OntologyChecker` created in this file

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+    return pd.isnull(v) or (hasattr(v, "__len__") and len(v) == 0)
+
+
+def _getattr_anndata(adata: anndata.AnnData, attr: str = None):
```
it does not use `self` at all, pycharm was telling me it's not best practice

atolopko-czi replied:
if made into a class method, then `self.adata` could be accessed. alternately, can use Python's `@staticmethod` annotation. quite minor!

pablo-gar replied:
Fixed in f68682ff4c918f36b5fae8f4fab1d2b98201caf4

brianraymor commented:
| ERROR: Could not infer organism from feature ID 'AURKAIP1' in 'var', make sure it is a valid ID

I agree. If the _heuristics_ fail to narrow down the right GTF to query, then it's simply an invalid identifier.

brianraymor commented:
> ERROR: 'ERCC3' is not a valid feature ID in 'var'

YAML check for feature_biotype includes:

                        # If feature id start with ERCC then only allowed for "spike-in"
                        rule: "index.str.startswith('ERCC')"

pablo-gar commented:
> > ERROR: 'ERCC3' is not a valid feature ID in 'var'
> > ERROR: Column 'feature_biotype' in dataframe 'var' contains invalid values '['gene']'. Values must be one of ['spike-in'].
> 
> YAML check for feature_biotype includes:
> 
> ```
>                         # If feature id start with ERCC then only allowed for "spike-in"
>                         rule: "index.str.startswith('ERCC')"
> ```
> 
> which should be 'ERCC-' ?

Good catch
fixed in 85b529c0c1022783d66a5760c7f8c8309d933e37

@jychien This should produce better error logging

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        # ATAC - raw layer not required
+        self.validator.adata.obs["assay_ontology_term_id"] = "EFO:0010891"
+        self.validator.validate_adata()
+        print("HOLAaaa", self.validator.errors)
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        )
+
+        # ATAC - raw layer not required
+        self.validator.adata.obs["assay_ontology_term_id"] = "EFO:0010891"
```
379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            "tissue_ontology_term_id",
+        ]
+
+        for column in columns:
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        Terms documented as obsolete in an ontology MUST NOT be used. For example, EFO:0009310
+        for obsolete_10x v2 was marked as obsolete in EFO version 3.31.0 and replaced by
+        EFO:0009899 for 10x 3' v2.
+        """
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_assay_ontology_term_id(self):
+
+        """
+        assay_ontology_term_id categorical with str categories. categorical with str categories.
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+class TestValidAnndata(unittest.TestCase):
+
+    """
+    Tests a valid anndata. Most other tests below modify this anndata object and test for failure cases
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            ],
+        )
+
+    def test_raw_existance(self):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            ],
+        )
+
+    def test_development_stage_ontology_term_id(self):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        self.validator.adata.obs["organism_ontology_term_id"][0] = "NCBITaxon:10090"
+        self.validator.adata.obs["development_stage_ontology_term_id"][
+            0
+        ] = "MmusDv:0000003"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        # this MUST be either a HANCESTRO term or "unknown" if unavailable.
+        self.validator.adata.obs["organism_ontology_term_id"][0] = "EFO:0000001"
+        self.validator.adata.obs["development_stage_ontology_term_id"][0] = "unknown"
+        self.validator.adata.obs["ethnicity_ontology_term_id"][0] = "na"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            ],
+        )
+
+    def test_tissue_ontology_term_id(self):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/fixtures/examples_validate.py:
```diff
@@ -0,0 +1,267 @@
+)
+
+# invalid obs, all fields are designed to fail
+bad_obs = pd.DataFrame(
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+def _getattr_anndata(adata: anndata.AnnData, attr: str = None):
+
+    """
+    same as getattr but handles the special case of "raw.var" for an anndata.Anddata object
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/fixtures/examples_validate.py:
```diff
@@ -0,0 +1,267 @@
+)
+
+# Anndata with no obs nor var
+adata_empty = anndata.AnnData(X=sparse.csr_matrix(X), uns=good_uns, obsm=good_obsm)
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        if column_def.get("type") == "bool":
+            if not column.dtype == bool:
+                self.errors.append(
+                    f"Column '{column_name}' in dataframe '{df_name}' must be boolean not '{column.dtype.name}'"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        if column_def.get("type") == "categorical":
+            if not column.dtype.name == "category":
+                self.errors.append(
+                    f"Column '{column_name}' in dataframe '{df_name}' must be categorical not {column.dtype.name}"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        """
+
+        self.validator.adata.raw = self.validator.adata
+        self.validator.adata.uns["X_normalization"] = "CPM"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar replied:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        """
+
+        # If the feature_biotype is "gene" then this MUST be an ENSEMBL term.
+        # First not and ENSEMBL ID
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_required_fields_schema_version(self):
+
+        """
+        Curators MUST annotate the following keys and values in uns: schema_version, title, X_normalization
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: Schema version '1.0.0' is not supported. Validation "
+                "cannot be performed."
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: 'cell_type_ontology_term_id' in 'uns['batch_condition']' "
+                "is not valid, it must be a list or numpy array"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+    def test_X_approximate_distribution_is_valid(self):
+
+        """
+        X_approximate_distribution str. The value MUST be "count" [...] or "normal"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        )
+
+
+class TestAddingLabels(unittest.TestCase):
```
I will opt to keep it here as it is directly referenced from the schema doc

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+
+    """
+    Tests the addition of labels from IDs based on schema specification. The test is done by comparing manually
+    created dataframes (positive control) agains the ones produced by the validator
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,215 @@
+        self.assertTrue(self.validator.errors)
+
+
+class TestH5adValidation(unittest.TestCase):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+            ],
+        )
+
+    def test_assay_ontology_term_id(self):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+
+            # Get array without zeros
+            if isinstance(x, ndarray):
+                non_zeroes_index = nonzero(x)
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        )
+
+
+class TestVar(unittest.TestCase):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_validate.py:
```diff
@@ -0,0 +1,215 @@
+            self.column_name
+        ]["curie_constraints"]
+
+    def test_schema_defintion(self):
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+                )
+            if "ontologies" not in column_def["curie_constraints"]:
+                raise ValueError(
+                    f"allowed 'ontolgies' must be specified under 'curie constraints' for '{column_name}'"
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+
+        success = validator.is_valid & writer.was_writing_successful
+
+    return success
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+    if not validator.is_valid:
+        return False
+
+    success = False
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/cli.py:
```diff
@@ -8,14 +10,34 @@
 
 
-schema_cli.add_command(placeholder)
+schema_cli.add_command(schema_validate)
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+        return getattr(adata, attr)
+
+
+def _get_schema_file_path(version: str) -> str:
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/tests/test_schema_compliance.py:
```diff
@@ -0,0 +1,1006 @@
+        this MUST be the most accurate HsapDv term.
+        If organism_ontolology_term_id is "NCBITaxon:10090" for Mus musculus,
+        this MUST be the most accurate MmusDv term
+        All other it MUST be children of UBERON:0000105 and not UBERON:0000071
```
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+ONTOLOGY_CHECKER = ontology.OntologyChecker()
+
+
+def _is_null(v):
```
https://github.com/chanzuckerberg/single-cell-curation/pull/116#discussion_r700366742

pablo-gar commented on cellxgene_schema_cli/cellxgene_schema/validate.py:
```diff
@@ -0,0 +1,1588 @@
+            else:
+                self.schema_def = _get_schema_definition(version)
+
+    def _get_component_def(self, component: str) -> dict:
```
> Hi @pablo-gar, I'm still going through test cases for X and raw.X, but I thought I'd list these initial issues, as one of the error messages is coming up on more than 1 occasion (which was the one I had slacked you about). In general, I'm am creating test cases where I am **not** assuming that 'X_normalization' was filled accurately, so the validator should flag datasets that are missing the raw layer when applicable:
> 
> * **X/raw.X**: ATAC-seq dataset with only normalized in X (should be okay), but results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF180GFI_v4.h5ad)
> * **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, and X_normalization = 'LogNormalization' results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF751GQY_v4_2.h5ad)
> * **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, X_normalization = none results in no logging error (test file = LATDF751GQY_v4_1.h5ad)
> * **X/raw.X**: have X and raw.X be different shapes results in no logging error (test file = LATDF516GAW_v4_2.h5ad)

@jychien all these should be fixed in the latest commit

pablo-gar commented:
> Hi @pablo-gar, I'm still going through test cases for X and raw.X, but I thought I'd list these initial issues, as one of the error messages is coming up on more than 1 occasion (which was the one I had slacked you about). In general, I'm am creating test cases where I am **not** assuming that 'X_normalization' was filled accurately, so the validator should flag datasets that are missing the raw layer when applicable:
> 
> * **X/raw.X**: ATAC-seq dataset with only normalized in X (should be okay), but results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF180GFI_v4.h5ad)
> * **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, and X_normalization = 'LogNormalization' results in `validate.py:925: RuntimeWarning: invalid value encountered in remainder`. (test file = LATDF751GQY_v4_2.h5ad)
> * **X/raw.X**: RNA-seq dataset without raw matrix, only normalized in X, X_normalization = none results in no logging error (test file = LATDF751GQY_v4_1.h5ad)
> * **X/raw.X**: have X and raw.X be different shapes results in no logging error (test file = LATDF516GAW_v4_2.h5ad)

@jychien all these should be fixed in the latest commit

jychien commented:
Thanks, @pablo-gar! Everything checks except that I am still getting the following error when validating a file that erroneously has gene symbols as index. Sample file is `180bff9c-c8a5-4539-b13b-ddbc00d643e6_06b91002-4d3d-4d2e-8484-20c3b31e232c.h5ad`
- ERROR: Column 'feature_biotype' in dataframe 'var' contains invalid values '['gene']'. Values must be one of ['spike-in'].

I've replicated some of the issues that we've found in migration, and would be great if the validator can log these types of errors/issues:
- RNA-seq dataset with normalized matrix in raw.X and raw in X. Test file is LATDF257RBX_v4_filtered.h5ad.
- RNA-seq dataset with normalized matrix in raw.X and normalized in X. Test file is LATDF257RBX_v4_2_filtered.h5ad
- RNA-seq dataset with raw matrix in raw.X and raw in X.  Would be great to flag for redundancy. Test file is LATDF257RBX_v4_2_filtered.h5ad
